### DISCUSSION TITLE
rag api: "POST /embed HTTP/1.1" 404 Not Found

### DISCUSSION THREAD

**Discussion by ajkost** [danny-avila/LibreChat#2162] [Troubleshooting] [answered]
Moved to discussion from [PR](https://github.com/danny-avila/LibreChat/pull/2143#issuecomment-2009896107) since I am facing the same issue. Thank you for looking into it.

I use `text-embedding-ada-002` on azure

logs from rag API
```
2024-03-22 15:38:50,763 - root - INFO - Initialized embeddings of type: <class 'langchain_openai.embeddings.azure.AzureOpenAIEmbeddings'>
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2024-03-22 15:38:51,200 - root - DEBUG - Created index 'idx_langchain_pg_embedding_custom_id' on 'langchain_pg_embedding(custom_id)'
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
2024-03-22 15:39:05,986 - root - WARNING - JWT_SECRET not found in environment variables
INFO:     127.0.0.6:41003 - "POST /embed HTTP/1.1" 404 Not Found
```

when trying to upload file in librechat
![image](https://github.com/danny-avila/LibreChat/assets/10452148/e05e4309-6a08-45e2-95f0-482804febb9c)

**Comment by danny-avila** [comment]
Currently, the RAG API needs to be started from the same directory as LibreChat.

For convenience, there is a new compose file, `rag.yml`

from LibreChat directory:

```bash
docker compose -f ./rag.yml pull
docker compose -f ./rag.yml up
```

Users have also had issues because they weren't using the latest docker images, so make sure you either build it locally or pull the latest librechat image (if using docker)

```bash
docker compose pull
docker compose up
```

> the RAG API needs to be started from the same directory as LibreChat.


This is a temporary method, and I'm working on replacing this way since it's not robust and was just to prototype the initial implementation.

**Reply by danny-avila** [reply]
Also please post LibreChat logs when you encounter the issue.

Following my instructions, I'm not able to reproduce the issue: 
![image](https://github.com/danny-avila/LibreChat/assets/110412045/f1768f26-9d95-44d1-9af1-541aa168d989)

**Reply by danny-avila** [reply]
![image](https://github.com/danny-avila/LibreChat/assets/110412045/433cbdd2-22b1-42a8-a439-040a17b59fe4)

**Comment by ajkost** [comment]
Thanks for the response. I am using kubernetes to run both librechat and rag_api separately, within a single namespace. For librechat I use `ghcr.io/danny-avila/librechat-dev-api:latest` so I assume it is the latest image and in deployment description I can see it was pulling the new image.

Logs from librechat
```
2024-03-22 16:03:27 error: Request failed with status code 404
2024-03-22 16:03:27 error: [/files] Error processing file: AxiosError: Request failed with status code 404
```

> the RAG API needs to be started from the same directory as LibreChat.

why so? what are the dependencies?

**Reply by danny-avila** [reply, marked as answer]
> why so? what are the dependencies?

Well, the RAG API is setup to be able to run local embeddings as well, making for a 6 GB image when installed, so I don't want to surprise anyone with the hefty download integrating this in the normal setup.

The main reason however, is that I'm still prototyping this, and the setup will change to allow for remote deployments.

Right now the only real dependency that is breaking this for you is not sharing volume storage with LibreChat.

**Reply by ajkost** [reply]
that's something I should be able to resolve by mounting the same storage to both. Is it about sharing `/app/client/public/` folder from librechat with upload folder in rag_api (whatever I configured as `RAG_UPLOAD_DIR` env variable)?

edit: or `/app/uploads` from librechat?

**Reply by danny-avila** [reply]
`rag.yml` compose file mounts volumes like this:

```yaml
    volumes:
      - ./uploads/temp:/app/uploads/temp
```

librechat `docker-compose.yml`

```
    volumes:
      - ./.env:/app/.env
      - ./images:/app/client/public/images
      - ./uploads:/app/uploads # <--- this is the relevant one for rag
      - ./logs:/app/api/logs
```

**Reply by danny-avila** [reply]
this was just a quick solution before needing to implement a different file processing strategy in the librechat backend, which I'm working on now.

**Reply by ajkost** [reply]
great, thanks, I will try this :+1: just to see how it works

**Reply by ajkost** [reply]
works like a charm! thanks again :tada:

**Reply by danny-avila** [reply]
sweet!

**Reply by spencerwongfeilong** [reply]
Hello i saw that if combining openai embedding with claude3 response, the results would not be as good as having openai embedding with gpt4. I was wandering if that is something that librechat can improve on? 


The response below shows it. But no issue with gpt4. 
![Screenshot_20240330_163112_Chrome](https://github.com/danny-avila/LibreChat/assets/127573355/acb9eafa-5a21-4670-8a6a-929c97b88265)

**Reply by danny-avila** [reply]
Hi @spencerwongfeilong 

This is not necessarily true. With LibreChat, what will affect the quality of the response the most is your prompting more than anything. The generative AI never interacts directly with the embeddings so it won't matter which AI you have selected (other than their inherent intelligence).

For example, with the exact same prompt `what is meant by the use of the word "attention" in this paper?`, the AI receive the same chunks of the document, and which responds better is both subjective and a crap-shoot.

![image](https://github.com/danny-avila/LibreChat/assets/110412045/b30718e6-10a4-4918-9370-4474fe1b638c)

**Reply by spencerwongfeilong** [reply]
Hi @danny-avila 

Thanks for the reply. Perhaps the screenshots below is a better representation. 

Its only only pertaining to this document. I have other documents which also present the same issue. 

By claude3
![Screenshot_20240330_200708_Chrome](https://github.com/danny-avila/LibreChat/assets/127573355/6a4f1826-7ab5-471c-af4e-059d696a67bb)

By gpt4
![Screenshot_20240330_200700_Chrome](https://github.com/danny-avila/LibreChat/assets/127573355/e6068622-7305-4579-a473-d16078350ed7)

**Reply by danny-avila** [reply]
In my opinion, Claude is giving the better response, because it's not getting enough information to summarize the full book. You would have to load the full book into context for it to give a real summary

**Reply by danny-avila** [reply]
You can actually enable this (very expensive) feature with env var `RAG_USE_FULL_CONTEXT=true`, however depending on the book size it may or may not fit

**Reply by spencerwongfeilong** [reply]
Hello. I believed the purpose of embedding the document was to find parts that closely match the query, and then create a response based on those parts.

This is another question with RAG in play showing the response for Claude 3 and gpt4 respectively. 

Tell me about chapter 14
Claude 3
Unfortunately, the context provided does not contain the full text of Chapter 14 titled "Wild Child." The excerpt given is only a small portion of that chapter, and does not provide enough information for me to summarize the chapter in detail.

The excerpt describes the author's experiences as a junior in high school, attending a party where she danced and interacted with a senior from another school. It mentions some flirtatious and intimate moments between the author and this other student. However, without the full context of the chapter, I do not have enough information to give a comprehensive summary.

I apologize that I cannot provide a more detailed overview of Chapter 14 based on the limited information available to me. If you are able to share the complete text of that chapter, I would be happy to review it and provide a more thorough summary. Please let me know if you have any other questions I could try to assist with.

Tell me about chapter 14
Gpt4
Chapter 14 of "Dear Girls" by Ali Wong is titled "Wild Child." In this chapter, Wong opens up about her rebellious teenage years, reflecting on her tumultuous relationship with her mother during that time. She candidly shares stories that embody the spirit of a teenager pushing boundaries and exploring her identity in sometimes questionable ways.

One poignant anecdote Wong shares is from when she was a junior in high school. She attended a pajama-themed party, where she engaged in a risqué dance with a senior from another high school, a much cooler one that was focused on arts and unconventional learning. Despite the party's theme, the young man's attire—an unthemed combination of a big, puffy, black North Face vest over a flannel shirt with cargo pants—struck a chord with Wong. Their interaction was bold and somewhat reckless, embodying Wong's "wild child" phase. They shared an intense moment of youthful indiscretion, dancing to Tupac's "California Love" and sharing a forty-ounce bottle of Mickey's malt liquor.

This chapter is significant as it showcases Wong's vulnerability and the complexity of her teenage years. She reflects on these memories with a mixture of humor and introspection, acknowledging the potential karma that awaits her as the mother of daughters who might one day explore their boundaries much like she did. Wong's recounting of her youth is not just a trip down memory lane; it's an honest reflection on the challenges of growing up and the impact of those formative years on her relationship with her mother and her perspective on life.

"Wild Child" is a chapter that resonates with readers for its relatable recounting of teenage rebellion and the messy, often complicated journey towards adulthood. Wong uses her unique voice to narrate these experiences, blending humor with heartfelt moments, ultimately offering a glimpse into the origins of her resilience and comedic outlook on life.

**Reply by spencerwongfeilong** [reply]
This is another book, twice the size of the previous book and Claude 3 and gpt4 were able to summarise. 

![Screenshot_20240330_204418_Chrome](https://github.com/danny-avila/LibreChat/assets/127573355/8cfd9797-4cb4-4188-9728-7f055e130081)

**Reply by spencerwongfeilong** [reply]
Hello @danny-avila 

It seems the problem here is related to chunk size. I set it to 1024 and Claude 3 was able to respond accordingly. Thank you.
